pkg/drawio/layout.go

```diff
@@ -0,0 +1,298 @@
+package drawio
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// layoutS is the main struct for layouting the tree nodes
+// overview to the layout algorithm:
+// the input to the layout algorithm is the tree itself. the output is the geometry for each node in the drawio (x, y, height, wight)
+// the steps:
+// 1. create a 2D matrix  - for each subnet icon, it set the location in the matrix
+// 2. set the locations of the SG in the matrix, according to the locations of the icons
+// 3. add squares borders - resizing the matrix. adding rows and columns to the matrix, to be used as the borders of all squares
+// 4. set the locations of all the squares in the matrix
+// 5. set the locations of all the non subnet icons
+// 6. set the geometry for each node in the drawio
+
+// the implementation of the matrix is the layoutMatrix struct.
+// layoutMatrix holds a list of column and a list of rows.
+// the layoutMatrix does not hold the treeNodes in it.
+// instead, each node holds pointers to its rows and columns (square can be in more than one rows/column).
+// alongside these pointers, the treNode holds the position inside the Matrix cell ( see struct Location)
+// that alow to add/remove rows/column without updating the treeNodes.
+
+const (
+	borderWidth  = 40
+	subnetWidth  = 8 * 40
+	subnetHeight = 6 * 40
+	iconSize     = 60
+	iconSpace    = 4 * 40
+
+	fipXOffset = -60
+	fipYOffset = 30
+
+	vsiXOffset  = 30
+	vsiYOffset  = -10
+	vsiIconSize = 40
+
+	// network -> vpc -> zone -> subnets
+	networkToSubnetDepth = 3
+)
+
+type layoutS struct {
+	network SquareTreeNodeInterface
+	matrix  *layoutMatrix
+}
+
+func newLayout(network SquareTreeNodeInterface) *layoutS {
+	ly := &layoutS{network: network, matrix: newLayoutMatrix()}
+	return ly
+}
+
+func (ly *layoutS) layout() {
+	// main layout algorithm:
+	// 1. create a 2D matrix  - for each subnet icon, it set the location in the matrix
+	ly.layoutSubnetsIcons()
+	ly.matrix.removeUnusedLayers()
+	// 2. set the locations of the SG in the matrix, according to the locations of the icons
+	ly.setSGLocations()
+	// 3. add squares borders - resizing the matrix. adding rows and columns to the matrix, to be used as the borders of all squares
+	ly.addAllBorderLayers()
+	// 4. set the locations of all the squares in the matrix
+	ly.setSquaresLocations()
+	ly.matrix.removeUnusedLayers()
+	// 5. set the locations of all the non-subnet icons
+	ly.setIconsLocations()
+	// 6. set the geometry for each node in the drawio
+	ly.matrix.setLayersDistance()
+	ly.setGeometries()
+}
+
+// ///////////////////////////////////////////////////////////////
+// layoutSubnetsIcons() implements a simple north-south east-west layouting:
+// 1. vpcs are next to each others
+// 2. zones are next to each others
+// 3. subnets a above/below each other
+// 4. cell can hold at most two icons
+// 5. only icons with the same sg can share a cell
+
```

Comment:
```suggestion
```
Go automatic documentation requires no spaces between comments and function

Reply:
done